src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

```diff
@@ -1494,78 +1496,137 @@ private string[] SplitWithPostProcessing(ReadOnlySpan<int> sepList, ReadOnlySpan
-                            sepListBuilder.Append(i);
-                        }
+                        sepListBuilder.Append(i);
                     }
-                    break;
-                case 2:
-                    sep0 = separators[0];
-                    sep1 = separators[1];
+                }
+            }
+
+            // Handle > 3 separators with a probabilistic map, ala IndexOfAny.
+            // This optimizes for chars being unlikely to match a separator.
+            else
+            {
+                unsafe
+                {
+                    ProbabilisticMap map = default;
+                    uint* charMap = (uint*)&map;
+                    InitializeProbabilisticMap(charMap, separators);
+
                     for (int i = 0; i < Length; i++)
                     {
                         char c = this[i];
-                        if (c == sep0 || c == sep1)
+                        if (IsCharBitSet(charMap, (byte)c) && IsCharBitSet(charMap, (byte)(c >> 8)) &&
+                            separators.Contains(c))
                         {
                             sepListBuilder.Append(i);
                         }
                     }
-                    break;
-                case 3:
-                    sep0 = separators[0];
-                    sep1 = separators[1];
-                    sep2 = separators[2];
-                    for (int i = 0; i < Length; i++)
+                }
+            }
+        }
+
+        private void MakeSeparatorListVectorized(ref ValueListBuilder<int> sepListBuilder, char c, char c2, char c3)
+        {
+            // Redundant test so we won't prejit remainder of this method
+            // on platforms without SSE.
+            if (!Sse41.IsSupported)
+            {
+                throw new PlatformNotSupportedException();
+            }
+
+            // Constant that allows for the truncation of 16-bit (FFFF/0000) values within a register to 4-bit (F/0)
+            Vector128<byte> shuffleConstant = Vector128.Create(0x00, 0x02, 0x04, 0x06, 0x08, 0x0A, 0x0C, 0x0E, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF);
+
+            Vector128<ushort> v1 = Vector128.Create(c);
+            Vector128<ushort> v2 = Vector128.Create(c2);
+            Vector128<ushort> v3 = Vector128.Create(c3);
+
+            ref char c0 = ref MemoryMarshal.GetReference(this.AsSpan());
+            int cond = Length & -Vector128<ushort>.Count;
+            int i = 0;
+
+            for (; i < cond; i += Vector128<ushort>.Count)
+            {
+                Vector128<ushort> charVector = ReadVector(ref c0, i);
+                Vector128<ushort> cmp = Sse2.CompareEqual(charVector, v1);
+
+                cmp = Sse2.Or(Sse2.CompareEqual(charVector, v2), cmp);
+                cmp = Sse2.Or(Sse2.CompareEqual(charVector, v3), cmp);
+
+                if (Sse41.TestZ(cmp, cmp)) { continue; }
+
+                Vector128<byte> mask = Sse2.ShiftRightLogical(cmp.AsUInt64(), 4).AsByte();
```

Comment:
There are a lot of test failures showing up in the PR that aren't showing up elsewhere. That's making me think there is a bug in the logic here, but its not obvious to me what that might be.

The goal here is to identify all of the 8 entries that matched as efficiently as possible and the comparison is a `0` or `ushort.MaxValue` per element. It might therefore be simpler to avoid the shuffle and shifting altogether and just use `Sse2.MoveMask` which will return `2-bits` per element where the bits are `00` (no match) or `11` (match).

The logic would then become:
```csharp
int mask = Sse2.MoveMask(cmp.AsByte());

// or unroll, its just 8 comparisons
for (int idx = 0; idx < 8; idx++)
{
     if ((mask & 1) != 0)
    {
        sepListBuilder.Append(i + idx);
    }
    mask >>= 2;
}
```

It could also be more clever with `LeadingZeroCount` to "exit early" if there is only one entry

Reply:
I've benchmarked your solution with and without exiting early and it looked to be overall slower than the non vectorized implementation. I assume this is because we essentially iterate over each character in the loop you proposed, as well as do all the vectorized comparisons. I managed to track down an issue with the implementation, essentially I should have shifted the mask by 32 bits as opposed to 16, generating wrong indices for the separatorlist. I hope this will fix the test failures.

Reply:
I'd guess that's because the existing comparison does `1-8` comparisons, just up until `lowBits` is 0 while mine always does 8.
Is it more efficient if you instead have `for (int idx = 0; mask != 0; idx++)` so its effectively the same number of comparisons as your current for loop?

Reply:
It is faster by having the early exiting condition, however still slower than the implementation in PR.
(SSE32 being in PR, and SSE32Imp being your suggestion)
|           Method |            CorpusUri |       Mean |
|----------------- |--------------------- |-----------:|
|    SplitCsvSSE32 | 1 | 188.676 us |
| SplitCsvSSE32Imp | 1 | 213.116 us |
|     SplitCsvLoop | 1 | 393.167 us |
|    SplitCsvSSE32 |  2 |   2.725 us |
| SplitCsvSSE32Imp |  2 |   3.831 us |
|     SplitCsvLoop |  2 |   5.789 us |
|    SplitCsvSSE32 | 3 |  17.662 us |
| SplitCsvSSE32Imp | 3 |  19.966 us |
|     SplitCsvLoop | 3 |  24.894 us |
|    SplitCsvSSE32 | 4 |   3.605 us |
| SplitCsvSSE32Imp | 4 |   4.093 us |
|     SplitCsvLoop | 4 |  10.071 us |
|    SplitCsvSSE32 |5| 140.898 us |
| SplitCsvSSE32Imp | 5 | 187.643 us |
|     SplitCsvLoop | 5 | 196.353 us |

Reply:
Thanks for checking!

Reply:
Unsure what to make of the tests failing/not completing this time. But it seems like certain tests that didn't pass last run, passed now. Is there a way to kick them off again?